### PR TITLE
URL encode links when a redirector is set

### LINKED ISSUE #328: links should be urlencoded when using a dereferer
Currently, if you use a dereferer in shaarli, there is an issue with links containing a fragment (the part after #).

As the url is not encoded, the browser parse the fragment and does not send it to to the dereferer.

Example with dereferer.org and zerobin :

With url not encoded (we are redirected to https://www.zerobin.net/?b82c165d5e558015) :

http://dereferer.org/?https://www.zerobin.net/?b82c165d5e558015#pxU6FIkc+J/LJ/Wj7Sr+2ZCkJZwiU+kWCWoSlmVqBdQ=
 

With url encoded, we are correctly redirected :

http://dereferer.org/?https%3A//www.zerobin.net/%3Fb82c165d5e558015%23pxU6FIkc%2BJ/LJ/Wj7Sr%2B2ZCkJZwiU%2BkWCWoSlmVqBdQ%3D

### PR DESCRIPTION
**Pull request by ArthurHoaro**
Fixes #328

  * WARNING - template edit - new variable available : "real_url"
  Contains the final real url (redirected or any other change on original URL)

  * Doesn't affect RSS feeds.

### CODE REVIEW

File: tpl/linklist.html [hunk 1]
```diff
@@ -50,7 +50,13 @@
                         <input type="hidden" name="token" value="{$token}"><input type="hidden" name="delete_link"><input type="image" alt="Delete" src="images/delete_icon.png#" title="Delete" class="button_delete" onClick="return confirmDeleteLink();"></form>
                     </div>
                 {/if}
-                <span class="linktitle"><a href="{$redirector}{$value.url}">{$value.title}</a></span>
+                <span class="linktitle">
+                    {if="$redirector==''"}
+                        <a href="{$value.url}">{$value.title}</a>
+                    {else}
+                        <a href="{$redirector}{function="urlencode($value.url)"}">{$value.title}</a>
```
Comment (virtualtam) [inline tpl/linklist.html]: Used across templates:
```php
{$value.url|urlencode}
```

File: tpl/linklist.html [hunk 2]
```diff
@@ -50,7 +50,13 @@
                         <input type="hidden" name="token" value="{$token}"><input type="hidden" name="delete_link"><input type="image" alt="Delete" src="images/delete_icon.png#" title="Delete" class="button_delete" onClick="return confirmDeleteLink();"></form>
                     </div>
                 {/if}
-                <span class="linktitle"><a href="{$redirector}{$value.url}">{$value.title}</a></span>
+                <span class="linktitle">
+                    {if="$redirector==''"}
```
Comment (virtualtam) [inline tpl/linklist.html]: are there other places where it could be relevant to support using a redirector?

**Comment by ArthurHoaro** [comment]
PR rewritten.

**Comment by ArthurHoaro** [comment]
- [ ] Fix and add unit tests.

File: tests/LinkDBTest.php [hunk 3]
```diff
@@ -510,4 +510,27 @@ public function testFilterFullTextMixed()
...
+            $this->assertEquals($link['url'], $link['real_url']);
+        }
+    }
+
+    /**
+     * Test real_url with redirector.
+     */
+    public function testLinkRealUrlWithRedirector()
+    {
+        $GLOBALS['redirector'] = 'http://redirector.to?';
```
Comment (virtualtam) [inline tests/LinkDBTest.php]: globals and tests may not mix well :)

to ensure other tests won't be affected by this value, it should be unset in the class' `setUp()` method

File: application/LinkDB.php [hunk 4]
```diff
@@ -263,7 +263,13 @@ private function _readDB()
 
         // Escape links data
         foreach($this->_links as &$link) { 
-            sanitizeLink($link); 
+            sanitizeLink($link);
+            if (!empty($GLOBALS['redirector']) && !startsWith($link['url'], '?')) {
```
Comment (virtualtam) [inline application/LinkDB.php]: it'd be safer not to use globals (they're hard enough to get rid of)

The redirector could be injected as a parameter in the class' constructor (see related comment in the tests)

**Comment by ArthurHoaro** [comment]
> The redirector could be injected as a parameter in the class' constructor

Done. Thanks for the review!

File: application/LinkDB.php [hunk 5]
```diff
@@ -57,18 +57,25 @@ class LinkDB implements Iterator, Countable, ArrayAccess
...
+     * @param boolean $hidePublicLinks if true all links are private.
+     * @param string  $redirector      link redirector set in user settings.
      */
-    function __construct($datastore, $isLoggedIn, $hidePublicLinks)
+    function __construct($datastore, $isLoggedIn, $hidePublicLinks, $redirector = '')
     {
         $this->_datastore = $datastore;
         $this->_loggedIn = $isLoggedIn;
         $this->_hidePublicLinks = $hidePublicLinks;
+        $this->_redirector = $redirector;
```
Comment (virtualtam) [inline application/LinkDB.php]: optional check - ensure the validator URL is valid:
- http://stackoverflow.com/q/2058578
- http://phpro.org/tutorials/Filtering-Data-with-PHP.html#8

File: application/LinkDB.php [hunk 6]
```diff
@@ -263,7 +270,13 @@ private function _readDB()
 
         // Escape links data
         foreach($this->_links as &$link) { 
-            sanitizeLink($link); 
+            sanitizeLink($link);
+            if (!empty($this->_redirector) && !startsWith($link['url'], '?')) {
```
Comment (virtualtam) [inline application/LinkDB.php]: an additional comment to precise we're not applying the redirector to internal links (notes) would be welcome :)

File: application/LinkDB.php (hunk 5 above)
Comment (ArthurHoaro) [inline application/LinkDB.php]: IMHO, if we do this, it should be done when the redirector is set.

Comment (virtualtam) [inline application/LinkDB.php]: Yep, this could be added when checking Shaarli global options and/or as a tool

Comment (ArthurHoaro) [inline application/LinkDB.php]: #381 opened.

**Comment by ArthurHoaro** [comment]
PR updated.

File: index.php [hunk 7]
```diff
@@ -779,7 +765,13 @@ function showRSS()
         // If user wants permalinks first, put the final link in description
         if ($usepermalinks===true) $descriptionlink = '(<a href="'.$absurl.'">Link</a>)';
         if (strlen($link['description'])>0) $descriptionlink = '<br>'.$descriptionlink;
-        echo '<description><![CDATA['.nl2br(keepMultipleSpaces(text2clickable($link['description']))).$descriptionlink.']]></description>'."\n</item>\n";
+        echo '<description><![CDATA['.
+            nl2br(
```
Comment (virtualtam) [inline index.php]: the method call seems to always be
```php
nl2br(keepMultipleSpaces(text2clickable($stuff, $redirector)))
```

could we introduce an helper class & method, e.g. `application/feed/Feed.php` / `Feed.formatLinkDescription($link, $redirector)`?

File: application/Utils.php [hunk 8]
```diff
@@ -168,3 +168,43 @@ function is_session_id_valid($sessionId)
...
+
+/**
+ * This function inserts &nbsp; where relevant so that multiple spaces are properly displayed in HTML
+ * even in the absence of <pre>  (This is used in description to keep text formatting).
+ *
+ * @param string $text input text.
+ *
+ * @return string formatted text.
+ */
+function keepMultipleSpaces($text)
```
Comment (virtualtam) [inline application/Utils.php]: convention for public functions: `keep_multiples_spaces`
maybe `space2nbsp` would best describe what it does?

**Comment by virtualtam** [comment]
Thanks! I've added a couple minor comments regarding the code structure

File: index.php (hunk 7 above)
Comment (ArthurHoaro) [inline index.php]: It doesn't only apply to feeds, but yes I agree with the idea.

**Comment by ArthurHoaro** [comment]
Thanks for your review. 

Update:

 * keepMultipleSpaces renamed to space2nbsp
 * space2nbsp improved to handle single space at line beginning
 * links in text description aren't 'nofollow' anymore*
 * new function `format_description` that's wrap previous functions and `nl2br()`.

[*] Other links aren't. I don't see why description links should be.

File: application/Utils.php [hunk 9]
```diff
@@ -168,3 +168,56 @@ function is_session_id_valid($sessionId)
...
+ * @return string formatted text.
+ */
+function space2nbsp($text)
+{
+    return preg_replace('/(^| ) /m', '$1&nbsp;', $text);
+}
+
+/**
+ * Format Shaarli's description
+ * TODO: Move me to ApplicationUtils when it's ready.
```
Comment (virtualtam) [inline application/Utils.php]: That's noted ;-)

**Comment by virtualtam** [comment]
+1, just needs to be rebased to take into account changes on `LinkDB()` calls

**Comment by ArthurHoaro** [comment]
Rebased.

**Comment by virtualtam** [comment]
Thanks :)

**Pull request merged**